DllImportGenerator/DllImportGenerator/TypePositionInfo.cs

```diff
@@ -307,21 +307,33 @@ static NativeMarshallingAttributeInfo CreateNativeMarshallingInfo(ITypeSymbol ty
                     NativeTypePinnable: ManualTypeMarshallingHelper.FindGetPinnableReference(nativeType) is not null);
             }
 
-            static bool TryCreateTypeBasedMarshallingInfo(ITypeSymbol type, DefaultMarshallingInfo defaultInfo, Compilation compilation, GeneratorDiagnostics diagnostics, out MarshallingInfo marshallingInfo)
+            static bool TryCreateTypeBasedMarshallingInfo(ITypeSymbol type, DefaultMarshallingInfo defaultInfo, Compilation compilation, GeneratorDiagnostics diagnostics, INamedTypeSymbol scopeSymbol, out MarshallingInfo marshallingInfo)
             {
                 var conversion = compilation.ClassifyCommonConversion(type, compilation.GetTypeByMetadataName(TypeNames.System_Runtime_InteropServices_SafeHandle)!);
                 if (conversion.Exists 
                     && conversion.IsImplicit 
                     && conversion.IsReference 
                     && !type.IsAbstract)
                 {
-                    marshallingInfo = new SafeHandleMarshallingInfo();
+                    bool hasAccessibleDefaultConstructor = false;
+                    if (type is INamedTypeSymbol named && named.InstanceConstructors.Length > 0)
+                    {
+                        foreach (var ctor in named.InstanceConstructors)
+                        {
+                            if (ctor.Parameters.Length == 0)
+                            {
+                                hasAccessibleDefaultConstructor = compilation.IsSymbolAccessibleWithin(ctor, scopeSymbol);
```

Comment:
Is `IsSymbolAccessibleWithin` just to handle cases where it is not `public` but still accessible? If so, would it be reasonable to just say it needs to be `public` (per the guidance that will be updated) for the generator to use the nicer generated code?

Reply:
We could do that, but I'd prefer to key this on "use the constructor directly if we can, otherwise use Activator.CreateInstance" as compared to "use the constructor directly only if the constructor is explicitly public" since it's easy to forget to mark an internal type's constructor as public (and a common pattern I saw in dotnet/runtime was to mark an internal type's constructor as internal).

Reply:
Using `IsSymbolAccessibleWithin` also allows SafeHandle subclasses to use static factory methods that call into a generated P/Invoke to create the handle to keep the default constructor private and still get the perf benefits of using the constructor directly.

The Activator.CreateInstance path is more a workaround to support the old pattern, at least in my mind.

Reply:
>[...] allows SafeHandle subclasses to use static factory methods that call into a generated P/Invoke to create the handle [...]

This is a nice use case example. I like @elinor-fung suggestion but as long as we address this API prior to supporting extensions I am "good" with it as is.

Reply:
Really nice use case. I agree we want to support that. Can we actually add an integration test that does that?

Reply:
Added an integration test. In the process found a small bug in how we generate the containing class declarations, so I've fixed that as well.